dm-manager/pkg/device/device.go

```diff
@@ -0,0 +1,475 @@
+	}
+
+	log.Debug().Msgf("desired state is %v[%v], current state is %v[%v] for %v",
+		invHost.GetDesiredAmtState().String(), invHost.GetDesiredPowerState().String(),
+		invHost.GetCurrentAmtState().String(), invHost.GetCurrentPowerState().String(), request.ID)
+
+	switch {
+	case invHost.GetCurrentAmtState() == computev1.AmtState_AMT_STATE_PROVISIONED &&
+		invHost.GetDesiredPowerState() == invHost.GetCurrentPowerState():
+		return dc.syncPowerStatus(ctx, request, invHost)
+
+	case invHost.GetCurrentAmtState() == computev1.AmtState_AMT_STATE_PROVISIONED &&
+		invHost.GetDesiredPowerState() != invHost.GetCurrentPowerState():
+		req, status, err := dc.handlePowerChange(ctx, request, invHost, invHost.GetDesiredPowerState())
+		if err != nil {
+			updateError := dc.updateHost(ctx, request.ID.GetTenantID(), invHost.GetResourceId(),
+				&fieldmaskpb.FieldMask{Paths: []string{
+					computev1.HostResourceFieldPowerStatus,
+					computev1.HostResourceFieldPowerStatusIndicator,
+				}}, &computev1.HostResource{
+					PowerStatus:          err.Error(),
+					PowerStatusIndicator: status,
+				})
+			if updateError != nil {
+				log.Err(updateError).Msgf("failed to update device status")
+				return request.Fail(updateError)
+			}
+		}
+
+		return req
+	default:
+		log.Debug().Msgf("device %v is in %v [%v]", request.ID, invHost.GetCurrentAmtState(), invHost.GetCurrentPowerState())
+	}
+
+	return request.Ack()
+}
+
+func (dc *Controller) syncPowerStatus(
+	ctx context.Context, request rec_v2.Request[ID], invHost *computev1.HostResource,
+) rec_v2.Directive[ID] {
+	currentPowerState, err := dc.MpsClient.GetApiV1AmtPowerStateGuidWithResponse(ctx, invHost.GetUuid())
+	if err != nil {
+		log.Err(err).Msgf("failed to check current power state for %v", invHost.GetUuid())
+		return request.Fail(err)
+	}
+
+	if currentPowerState.StatusCode() != http.StatusOK {
+		err = errors.Errorf("%v", string(currentPowerState.Body))
+		log.Err(err).
+			Msgf("expected to get 2XX, but got %v", currentPowerState.StatusCode())
+		return request.Fail(err)
+	}
+
+	//nolint: gosec // overflow is unlikely, correct values are <1000
+	powerStateCode := int32(*currentPowerState.JSON200.Powerstate)
+
+	if !contains(allowedPowerStates[invHost.GetDesiredPowerState()], powerStateCode) {
+		log.Info().Msgf("%v host desired state is %v, but current power state is %v",
+			invHost.GetUuid(), powerMapping[invHost.GetDesiredPowerState()], powerStateCode)
+
+		err = dc.updateHost(ctx, invHost.GetTenantId(), invHost.GetResourceId(),
+			&fieldmaskpb.FieldMask{Paths: []string{
+				computev1.HostResourceFieldCurrentPowerState,
+				computev1.HostResourceFieldPowerStatus,
+				computev1.HostResourceFieldPowerStatusIndicator,
+			}}, &computev1.HostResource{
+				PowerStatus:          "mismatch between desired and current power state",
+				PowerStatusIndicator: statusv1.StatusIndication_STATUS_INDICATION_IN_PROGRESS,
+				CurrentPowerState:    mpsPowerStateToInventoryPowerState[powerStateCode],
```

Comment:
and we should try to prevent unnecessary wirites (see my comment below line 376)